runtime/vm/jnicsup.cpp

```diff
@@ -498,30 +494,45 @@ UDATA JNICALL   pushArguments(J9VMThread *vmThread, J9Method* method, void *args
 				*(I_32*)(--sp) = (I_32)(jint)ARG(jint, i);	/* NOTE: jint may be larger than int, but jchar, jshort, etc, must be no larger than int */
 				break;
 #ifdef J9VM_INTERP_FLOAT_SUPPORT
-			case 'F':
+			case 'F': {
 				/* float type */
-				fltPtr = (jfloat*)--sp;
-				/* The Linux SH4 compiler needs the next two lines to be two steps.  If you combine them it fails to compile */
-				dbl = ARG(jdouble, f);
-				*fltPtr = (jfloat)dbl;
-				break;
+				jfloat *fltPtr = (jfloat*)--sp;
+				/* jfloat {aka float}' is promoted to 'double' when passed through '...' */
+				jdouble tempFloat = ARG(jdouble, f);
+				*fltPtr = (jfloat)tempFloat;
+			}
+			break;
 			case 'D':
 				/* double type */
-				dbl = ARG(jdouble, d);
-				lngOrDblPtr = (UDATA *) &dbl;
-				goto pushLongOrDouble;
+#ifdef J9VM_ENV_DATA64
+				sp -= 2;
+				*sp = (UDATA) ARG(jdouble, d);
+#else
+				{
+					jdouble tempDouble = ARG(jdouble, d);
+					UDATA * tempDblPtr = (UDATA *) &tempDouble;
+					sp -= 1;
+					*sp = *(tempDblPtr + 1);
+					sp -= 1;
+					*sp = *(tempDblPtr);
+				}
+#endif
+				break;
 #endif
 			case 'J':
 				/* long type */
-				lng = ARG(jlong, j);
-				lngOrDblPtr = (UDATA *) &lng;
-pushLongOrDouble:
 #ifdef J9VM_ENV_DATA64
-				--sp;
-				*--sp = *(lngOrDblPtr);
+				sp -= 2;
+				*sp = (UDATA) ARG(jlong, j);
```

Comment:
This code should work for both 32 and 64 bit.

Reply:
Rather, `*(U_64*)sp = ARG(...)`

Reply:
Likely a direct translation from builder, which could not deal with values larger than UDATA.